proc/groupby.go

```diff
@@ -340,63 +351,58 @@ func (g *GroupByAggregator) recordsForTable(table map[string]*GroupByRow) []*zso
 		keys = append(keys, k)
 	}
 	sort.Strings(keys)
-	var recs []*zson.Record
 
 	// We make multiple passes over the table, one for each unique type of
 	// vector or row keys, then in each pass we output as a batch all the
 	// records that have this unique type combo (i.e., the same descriptor).
 	// This allows variations in the type of the join-key to be tracked
 	// separately and thus generated each with their correct descriptor.
-	for _, typeCol := range g.typeCols {
-		for _, k := range keys {
-			row := table[k]
-			if !typeMatch(typeCol, row.rowKeys) {
-				continue
-			}
-			var zv zval.Encoding
-			if g.TimeBinDuration > 0 {
-				zv = zval.AppendValue(zv, []byte(row.ts.StringFloat()))
-			}
-			for _, rowkey := range row.rowKeys {
-				zv = zval.Append(zv, rowkey.Body, zeek.IsContainerType(rowkey.Type))
-			}
-			for _, red := range row.columns.Reducers {
-				// a reducer value is never a container
-				v := reducer.Result(red)
-				if zeek.IsContainer(v) {
-					panic("internal bug: reducer result cannot be a container!")
-				}
-				zv = v.Encode(zv)
+	n := len(g.keys) + len(g.reducerDefs)
+	if g.TimeBinDuration > 0 {
+		n++
+	}
+	scratchCols := make([]zeek.Column, 0, n)
+
+	var recs []*zson.Record
+	for _, k := range keys {
+		row := table[k]
+		var zv zval.Encoding
+		if g.TimeBinDuration > 0 {
+			zv = zval.AppendValue(zv, []byte(row.ts.StringFloat()))
+		}
+		zv = append(zv, row.keyvals...)
+		for _, red := range row.reducers.Reducers {
+			// a reducer value is never a container
+			v := reducer.Result(red)
+			if zeek.IsContainer(v) {
+				panic("internal bug: reducer result cannot be a container!")
 			}
-			d := g.lookupDescriptor(row)
-			r := zson.NewRecord(d, row.ts, zv)
-			recs = append(recs, r)
+			zv = v.Encode(zv)
 		}
+		d := g.lookupDescriptor(row, scratchCols)
```

Comment:
I don't understand how this works if we have to create multiple output descriptors, doesn't scratchCols keep getting appended to?

Reply:
scratchCols is zero-length in the caller and the slice value gets appended to in the callee which doesn't affect the caller's slice value.   It's just a way to avoid allocating a the slice every time lookupDescriptor is called.

Reply:
Hm, go arrays and slices are still unintuitive to me.  shrug...